2.3.x bug install date format mac os pr

devtom30 opened the pull request:
On Mac OS X, even if the LC_ALL variable is set to 'C', the install date format in system_profiler command output differs from one version to another. For instance, it's not the same format in version 10.8 and 10.7.
To avoid this problem, we use now the XML output of system_profiler command to retrieve the SOFTWARES section's data.

g-bougard commented:
Update is complex but I know @devtom30 heavily tested it. So merging it as this is an MacOS specific and expected fix, and it should not introduce regression anywhere else.

guillomovitch commented:
Parsing XML manually is still ugly, and lead to horrible code, such as extractApplicationsDataFromXmlLines() function: this is unreadable and unmaintainable. This part should be rewritten using XML::PP.

A version comparaison version function already exists in Tools, as compareVersion(), without the parsing part. It should be probably get renamed as compareVersionTokens(), and the cmpVersionNumbers() renamed as compareVersionStrings(), in the same namespace.

getSystemProfilerInfosXML is not declared in export list, is not documented, and its result is not cached. Is it on purpose ?

Lastly, I'd rather have a single getSystemProfilerInfos(), with a 'format' parameter (either "text" or "xml") instead of a distinct version, to enforce the fact it deserves the same purpose.

devtom30 commented:
Hello,

I agree with about horrible code. I did because of the XML output. The module XML::TreePP does not keep the order. But, as @g-bougard just pointed out to me, there is an option, depending on an additional core module, that could guarantee the elements order. I have to do some tests and I look if we can rewrite this horrible code !

devtom30 commented:
About comparison version functions, I understand well. The last introduced function can be generalized and renamed, and the other can be also renamed.
getSystemProfilerInfosXML : 
- not cached : it should be, we will do ti
- not exported : it should also but as you point out in your last sentence, it's true it's better to have a single function with a "format" parameter.

Thanks for your feedback !
